### PR TITLE
WIP: data simplification

### PR DESCRIPTION
**Pull request by erikbern**
We have all this shell scripts to download/prepare data. I tried to run them and they don't even run on my computer.

I rewrote some of the datasets (MNIST, Fashion-MNIST, SIFT, Glove) into a script that
1. Downloads the dataset
2. Splits into train/test
3. Prepares all the correct results (using BruteforceBLAS) and builds a HDF5 file.

Furthermore, I went through the rest of the code and 
1. Fixed Python3 compatibility
2. Use HDF5 for results as well

This ended up being a bit larger than I expected because I didn't really have a plan and a lot of stuff got pulled in. Have verified that it works.

@maumueller @ale-f curious to hear what you think

### CODE REVIEW

File: ann_benchmarks/results.py [hunk 1]
```diff
@@ -1,57 +1,55 @@
...
+        distances[i] = [d for n, d in ds]
+    f.close()
 
 def _get_leaf_paths(path):
     if os.path.isdir(path):
         for fragment in os.listdir(path):
             for i in _get_leaf_paths(os.path.join(path, fragment)):
                 yield i
-    elif os.path.isfile(path) and path.endswith(".json.gz"):
+    elif os.path.isfile(path) and path.endswith(".hdf5"):
```
[the remark this answers is not in the document]
Comment (erikbern) [reply ann_benchmarks/results.py]: 1. Performance reasons – HDF5 is much faster than json for numerical data (and also much smaller)
2. Already using HDF5 for the input data means some code (eg `metrics.py`) can compare two HDF5 files

File: data/prepare.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+    train = _load_mnist_vectors('fashion-mnist-train.gz')
+    test = _load_mnist_vectors('fashion-mnist-test.gz')
+    write_output(train, test, out_fn, 'euclidean')
+
+
+def random(out_fn, n_dims, n_samples, centers, distance):
+    X, _ = sklearn.datasets.make_blobs(n_samples=n_samples, n_features=n_dims, centers=centers, random_state=1)
+    X_train, X_test = sklearn.model_selection.train_test_split(X, test_size=0.1, random_state=1)
+    write_output(X_train, X_test, out_fn, distance)
+
```
[the remark this answers is not in the document]
Comment (erikbern) [reply data/prepare.py]: i glanced at your random code and it looked like something similar so i just replaced it with `make_blobs` – maybe it was vastly different?

**Comment by erikbern** [comment]
@maumueller I haven't ported all of the datasets – let me know what you want to do with:

* NYTimes, setsimilarity, hamming – these are static files on your server. Would prefer it if there's a reproducible build pipeline from the raw public data
* Webspam – it depends on "lib-annitu" which I don't know what it is :)
* Random – already left a comment about this

**Comment by erikbern** [comment]
Reran Glove-100 with Annoy on my laptop overnight

image

File: ann_benchmarks/plotting/metrics.py [hunk 3]
```diff
@@ -1,50 +1,48 @@
...
+def knn(dataset, run, epsilon=1e-10):
     total = 0
     actual = 0
-    for (query, max_distance, closest), [time, candidates] in results:
-        within = filter(lambda (index, distance): \
-                        distance <= (max_distance + epsilon), candidates)
-        total += len(closest)
+    count = int(run.attrs['candidates'])
+    for true_distances, found_distances in zip(dataset['distances'], run['distances']):
+        within = [d for d in found_distances if d <= true_distances[count] + epsilon]
```
[the remark this answers is not in the document]
Comment (erikbern) [reply ann_benchmarks/plotting/metrics.py]: i think you're right – should be `total += count`

**Comment by erikbern** [comment]
> @maumueller I haven't ported all of the datasets – let me know what you want to do with:
> 
> NYTimes, setsimilarity, hamming – these are static files on your server. Would prefer it if there's a reproducible build pipeline from the raw public data
> Webspam – it depends on "lib-annitu" which I don't know what it is :)
> Random – already left a comment about this

@maumueller ping

**Comment by erikbern** [comment]
going to merge this for now.

follow up PR will be to dockerize all algorithms and get the CI working again

**Comment by maumueller** [comment]
> @maumueller I haven't ported all of the datasets – let me know what you want to do with:
> NYTimes, setsimilarity, hamming – these are static files on your server. Would prefer it if there's a reproducible build pipeline from the raw public data
Webspam – it depends on "lib-annitu" which I don't know what it is :)
Random – already left a comment about this

- I'm going to run some experiments on the random dataset to compare how difficult they are.
- The source for the "set similarity" datasets is the "raw public data", I think. 
- I'll make a streamlined version for NYTimes and Webspam, I agree that this would be nice to have. 
- The hamming datasets: It takes quite a while to produce them, and the tools that are used in the process are far away from being nice to use. I'm not sure if this is really worth the effort (we are talking about something like two hours for a single dataset). 

Should I report on my results in here or start something new?

**Comment by erikbern** [comment]
> I'm going to run some experiments on the random dataset to compare how difficult they are.
> The source for the "set similarity" datasets is the "raw public data", I think.

I'll take a look

> I'll make a streamlined version for NYTimes and Webspam, I agree that this would be nice to have.

Ok, great

> The hamming datasets: It takes quite a while to produce them, and the tools that are used in the process are far away from being nice to use. I'm not sure if this is really worth the effort (we are talking about something like two hours for a single dataset).

It's fine that it takes a fair amount of time – once the data is built, I can upload it to S3 and I don't think anyone has to reproduce it. But it would be good if we stick to standard tools.

> Should I report on my results in here or start something new?

Can start a new pull request!

**Comment by maumueller** [comment]
Two more comments on the work done here: 

1. The plotting became horrible slow. I suspect this comes from the list slicing done in metrics.py. 
2. It broke createwebsite.py, and there are some methods missing that are used in there. So, I'm not sure how to fix it.

**Comment by erikbern** [comment]
> The plotting became horrible slow. I suspect this comes from the list slicing done in metrics.py.

ok, i'll take a look

> It broke createwebsite.py, and there are some methods missing that are used in there. So, I'm not sure how to fix it.

that's what unit tests are for :)

i'll try to run it once i'm done with the docker PR

**Pull request merged**